### PR TITLE
VS Code Extension for Parcel

### PR DESCRIPTION
**Pull request by thebriando**
# ↪️ Pull Request
## How it works
image
### Parcel LSP Reporter
`buildProgress` events report information during each Parcel phase and we translate these events into VS Code’s [Work Done Progress events](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#workDoneProgress).

`buildFailure` events return an array of Parcel `Diagnostic`'s, which we translate into VS Code [Diagnostics](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#:~:text=Range%3B%0A%7D-,Diagnostic,-Represents%20a%20diagnostic).

The reporter will also create a descriptor file that contains information about the Parcel process, and the translated diagnostics will be emitted through the IPC server.

### LSP Server

The LSP Server will watch for Parcel descriptors created by the reporter, establish an IPC client connection for each Parcel process, and report the diagnostics that are received from the reporter to each client.

## How to use it
https://github.com/parcel-bundler/parcel/blob/vscode-extension/packages/utils/parcelforvscode/README.md

## Todo
- [ ] Publish to VS Code Marketplace
- [ ] Hints
- [ ] Quick fix

## 💻 Examples

Build phase progress messages in VS Code
image

Parcel diagnostics in VS Code
image

## ✔️ PR Todo

- [ ] Added/updated unit tests for this change
- [ ] Filled out test instructions (In case there aren't any unit tests)
- [ ] Included links to related issues/PRs

### CODE REVIEW

**Comment by mischnic** [comment]
I think it would be nice to always show something in the status bus when the extension is connected to a Parcel instance (similar to the Flow extension, so like a static progress message)?

**Comment by parcel-benchmark** [comment]
## Benchmark Results
<details><summary>Kitchen Sink ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 2.36s | +372.00ms ⚠️ |
| Cached | 406.00ms | +29.00ms ⚠️ |

#### Cold Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/legacy/parcel.7cdb0fad.webp | 102.94kb | +0.00b | 389.00ms | +55.00ms ⚠️ |
| dist/legacy/parcel.7cdb0fad.webp | 102.94kb | +0.00b | 391.00ms | +56.00ms ⚠️ |
| dist/modern/parcel.7cdb0fad.webp | 102.94kb | +0.00b | 391.00ms | +55.00ms ⚠️ |
| dist/legacy/index.2c76ad23.js | 1.66kb | +0.00b | 712.00ms | +101.00ms ⚠️ |
| dist/legacy/index.8aaa89c9.js | 1.20kb | +0.00b | 712.00ms | +100.00ms ⚠️ |
| dist/modern/index.6be20f01.js | 1.13kb | +0.00b | 712.00ms | +100.00ms ⚠️ |
| dist/legacy/index.html | 826.00b | +0.00b | 904.00ms | +127.00ms ⚠️ |
| dist/modern/index.html | 749.00b | +0.00b | 903.00ms | +127.00ms ⚠️ |
| dist/legacy/index.b8ae99ba.css | 94.00b | +0.00b | 408.00ms | +64.00ms ⚠️ |
| dist/modern/index.31cedca9.css | 94.00b | +0.00b | 407.00ms | +62.00ms ⚠️ |

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/legacy/parcel.7cdb0fad.webp | 102.94kb | +0.00b | 357.00ms | +44.00ms ⚠️ |
| dist/legacy/parcel.7cdb0fad.webp | 102.94kb | +0.00b | 357.00ms | +44.00ms ⚠️ |
| dist/modern/parcel.7cdb0fad.webp | 102.94kb | +0.00b | 357.00ms | +43.00ms ⚠️ |
| dist/legacy/index.2c76ad23.js | 1.66kb | +0.00b | 668.00ms | +92.00ms ⚠️ |
| dist/legacy/index.8aaa89c9.js | 1.20kb | +0.00b | 667.00ms | +92.00ms ⚠️ |
| dist/modern/index.6be20f01.js | 1.13kb | +0.00b | 667.00ms | +92.00ms ⚠️ |
| dist/legacy/index.html | 826.00b | +0.00b | 819.00ms | +86.00ms ⚠️ |
| dist/modern/index.html | 749.00b | +0.00b | 818.00ms | +85.00ms ⚠️ |
| dist/legacy/index.b8ae99ba.css | 94.00b | +0.00b | 366.00ms | +43.00ms ⚠️ |
| dist/modern/index.31cedca9.css | 94.00b | +0.00b | 366.00ms | +43.00ms ⚠️ |

</p></details>

<details><summary>React HackerNews ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 11.68s | +239.00ms |
| Cached | 507.00ms | -7.00ms |

#### Cold Bundles

*No bundle changes detected.*

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/index.js | 486.39kb | +0.00b | 6.46s | +332.00ms ⚠️ |
| dist/PermalinkedComment.46b19af5.js | 4.21kb | +0.00b | 6.46s | +331.00ms ⚠️ |
| dist/UserProfile.f8cd7884.js | 1.57kb | +0.00b | 6.46s | +331.00ms ⚠️ |
| dist/NotFound.960ab92b.js | 429.00b | +0.00b | 6.46s | +331.00ms ⚠️ |
| dist/logo.c5bb83f1.png | 246.00b | +0.00b | 6.41s | +331.00ms ⚠️ |

</p></details>

<details><summary>AtlasKit Editor ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 2.00m | -4.18s |
| Cached | 2.97s | -361.00ms 🚀 |

#### Cold Bundles

*No bundle changes detected.*

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/index.1fa5b6e7.js | 3.33mb | +0.00b | 38.52s | -2.51s 🚀 |
| dist/EmojiPickerComponent.9d8e543c.js | 187.29kb | +0.00b | 38.54s | -2.53s 🚀 |
| dist/mobile-upload.b893855e.js | 68.22kb | +0.00b | 38.52s | -2.51s 🚀 |
| dist/mobile-upload.bf388fdf.js | 8.10kb | +0.00b | 38.52s | -2.51s 🚀 |
| dist/EmojiPickerComponent.16132f88.js | 5.54kb | +0.00b | 38.52s | -2.51s 🚀 |
| dist/ResourcedEmojiComponent.bef36b85.js | 2.68kb | +0.00b | 38.52s | -2.52s 🚀 |
| dist/date.728667d8.js | 2.10kb | +0.00b | 38.52s | -2.52s 🚀 |
| dist/images.0fe0f35c.js | 2.05kb | +0.00b | 38.52s | -2.52s 🚀 |
| dist/status.2160efcd.js | 1.82kb | +0.00b | 38.53s | -2.52s 🚀 |
| dist/workerHasher.0904bc5b.js | 1.72kb | +0.00b | 38.52s | -2.51s 🚀 |
| dist/code.414e2c00.js | 1.72kb | +0.00b | 38.52s | -2.52s 🚀 |
| dist/list-number.2be0cd6c.js | 1.62kb | +0.00b | 38.52s | -2.52s 🚀 |
| dist/link.1e33efbb.js | 1.43kb | +0.00b | 38.52s | -2.52s 🚀 |
| dist/emoji.5404dc09.js | 1.40kb | +0.00b | 38.52s | -2.52s 🚀 |
| dist/mention.85e501f2.js | 1.24kb | +0.00b | 38.53s | -2.51s 🚀 |
| dist/layout.dde5aa7a.js | 1.20kb | +0.00b | 38.52s | -2.52s 🚀 |
| dist/divider.3af42f3a.js | 1.20kb | +0.00b | 38.52s | -2.52s 🚀 |
| dist/action.633b1d3a.js | 1.17kb | +0.00b | 38.52s | -2.52s 🚀 |
| dist/heading1.216107b4.js | 1.17kb | +0.00b | 38.53s | -2.53s 🚀 |
| dist/list.1bd3e768.js | 1.14kb | +0.00b | 38.52s | -2.52s 🚀 |
| dist/quote.707178c1.js | 1.14kb | +0.00b | 38.53s | -2.52s 🚀 |
| dist/decision.5c187f01.js | 1.12kb | +0.00b | 38.52s | -2.52s 🚀 |
| dist/panel-warning.817f153e.js | 1.10kb | +0.00b | 38.53s | -2.52s 🚀 |
| dist/table.71e953fa.js | 1.08kb | +0.00b | 38.53s | -2.52s 🚀 |
| dist/panel.63b53aef.js | 1.02kb | +0.00b | 38.53s | -2.52s 🚀 |
| dist/panel-error.3c204bc9.js | 1019.00b | +0.00b | 38.53s | -2.51s 🚀 |
| dist/panel-success.bfed2890.js | 960.00b | +0.00b | 38.53s | -2.51s 🚀 |
| dist/panel-note.badaa0b2.js | 950.00b | +0.00b | 38.53s | -2.51s 🚀 |
| dist/simpleHasher.2e0de700.js | 742.00b | +0.00b | 38.52s | -2.51s 🚀 |

</p></details>

<details><summary>Three.js ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 7.64s | -1.06s 🚀 |
| Cached | 381.00ms | +20.00ms ⚠️ |

#### Cold Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/Three.js | 580.04kb | +0.00b | 5.35s | -890.00ms 🚀 |

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/Three.js | 580.04kb | +0.00b | 5.66s | -882.00ms 🚀 |

</p></details>

**[Click here to view a detailed benchmark overview.](https://benchmark.parceljs.org/benchmark/338271175083819603)**

**Review by devongovett — APPROVED** [review]
Do we wanna merge this and iterate or is this waiting for something?

**Comment by thebriando** [comment]
@devongovett Yeah we could probably merge this. I think I was just waiting to publish the extension to the Marketplace but I can do that in a separate task

File: packages/reporters/lsp-reporter/src/LspReporter.js [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+import invariant from 'assert';
+import path from 'path';
+import os from 'os';
+import fs from 'fs';
+// flowlint-next-line untyped-import:off
+import ps from 'ps-node';
+import {promisify} from 'util';
+// flowlint-next-line untyped-import:off
+import ipc from 'node-ipc';
+import {getProgressMessage} from '../../cli/src/utils';
```
Comment (mischnic) [inline packages/reporters/lsp-reporter/src/LspReporter.js]: This `Import for monorepo package '@parcel/reporter-cli' should be absolute` eslint error will definitely have to be fixed somehow

Comment (thebriando) [reply packages/reporters/lsp-reporter/src/LspReporter.js]: Hmm maybe we can move that to `@parcel/utils`? Or create a new package for reporter utils?

**Pull request merged**